test/src/sdk-harness/test_projects/call_frames/src/main.sw

```diff
@@ -21,7 +22,16 @@ impl CallFramesTest for Contract {
         first_param()
     }
 
-    fn get_second_param() -> u64 {
+    fn get_second_param_u64(arg0: u64) -> u64 {
         second_param()
     }
+
+    fn get_second_param_bool(arg0: bool) -> bool {
+        read(second_param())
```

Comment:
Is this read needed here? Bool and u64 should exhibit the same behaviour. I believe a read is only needed for multiple arguments or single arguments of reference types. You could test this out by covering the possible cases:
- Single copy type argument (u64 or bool). You already have tests for that but the bool probably needs an small update.
- Single reference type argument (b256 or struct for example). This is missing a test case.
- Multiple copy type arguments. You have a test for that but would be nice to throw in some reference types in there.

Reply:
since I changed second_param to return a u64 instead of T I think I'm unable to just return second_param() like I had before

Reply:
I'll do some experimenting

Reply:
Ohhh you're right. This is annoying because second_param is meant to give you the value itself if you only have a single copy type argument and so calling read on the result is not really correct I think. So what you want is a cast from u64 to bool. 

The fact that param2 sometimes contains a value and sometimes contains a pointer is what's causing the confusion here. So maybe we should go back to keeping second_param a generic and calling read inside it with <T> and see if this covers all the cases without having to call read anywhere else in the tests?

I'm sorry for the back and forth on this. This is what I get for trying to be helpful from my phone on a day off 😞

Reply:
All good I appreciate the help you've given me!  Enjoy your day off :)

Reply:
so should second_param() ever return a pointer or should it handle pointers itself?  That is should second_param() handle reading  such that the test functions don't have to.

Reply:
The cleanest implementation seems to be to have second_param return T and just call read inside it with <T> as a type argument. Then just calling second_param from the test with the same type arguments should give you the right result in all cases I believe.

Reply:
But we should document the meaning of T in this case and how to set it properly based on the arguments of the contract method.

Reply:
hmm I tried
```
pub fn second_param<T>() -> T {
    read::<T>(frame_ptr() + SECOND_PARAMETER_OFFSET)
}
```
and
```
fn get_second_param_multiple_params(arg0: bool, arg1: u64) -> (bool, u64) {
        second_param::<(bool, u64)>()
}
```
But it's not working as intended.  I'll take another look at it later.

Reply:
So I looked into this a bit and it seems that one solution is to implement `second_param` as follows:
```rust
/// Get the second parameter from the current call frame.
pub fn second_param<T>() -> T {
    if !is_reference_type::<T>() {
        read::<T>(frame_ptr() + SECOND_PARAMETER_OFFSET)
    }
    else {
        read::<T>(read::<u64>(frame_ptr() + SECOND_PARAMETER_OFFSET))
    }
}
```
Basically, if `T` is a reference type, then we need to read `param2` itself as a `u64` and then read from the pointer stored there. Otherwise, `param2` is just the value itself (this would be the case of single argument that is an integer or bool).

Another alternative is to implement 4 functions as follows:

- `fn first_param()`: gives back the raw first param as `u64` using a single `read()`.
- `fn second_param(0`: gives back the raw second param as `u64` using a single `read()`.
- `fn selector()`: returns the selector which happens to be the first parameter, so we just call `first_param()` in the body of `fn selector()`.
- `fn call_data()`: gives back the call data implemented as above.

What do you think? 
cc @nfurfaro @adlerjohn

Reply:
With this, we never need to call `read()` from the test.

Reply:
I definitely would prefer to not have to call `read()` from the test.
It sounds like the second alternative may unblock @matt-user for his work on a wrapper function for `call`?